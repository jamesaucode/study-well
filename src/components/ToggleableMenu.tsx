import React, { useState, useEffect } from "react";
import styled from "styled-components";
import Link from "next/link";
import Modal from "../components/Modal";
import Login from "../components/Login";

interface PropTypes {
  iconName: string;
  loggedIn: boolean;
}
interface MenuProps {
  expanded?: boolean;
}
const DropDownWrapper = styled.div`
  position: relative;
  display: inline-block;
  align-items: center;
  padding: 0 1rem;
  margin: 0 auto;
`;

const DropdownMenu = styled.div<MenuProps>`
  border: none;
  min-width: 160px;
  padding: 0.75rem 0;
  background: #8610f9;
  border-radius: 3px;
  box-shadow: 4px 4px 8px rgba(0, 0, 0, 0.1);
  display: ${props => (props.expanded ? "flex" : "none")};
  flex-direction: column;
  align-items: center;
  position: absolute;
  right: 0;
  z-index: 1;
`;

const LinkWrapper = styled.div`
  height: 2.5em;
  width: 100%;
  display: flex;
  align-items: center;
  &:hover {
    cursor: pointer;
    background: #a72ffb;
  }
`;
const StyledLink = styled.a`
  font-size: 0.9em;
  font-weight: 600;
  padding: 0 1rem;
  color: white;
  text-decoration: none;
`;
const ImportantLink = styled(StyledLink)`
  /* border: 1px solid #4285f4; */
  border: 1px solid #0f76fc;
  border-radius: 3px;
  /* color: #4285f4; */
  color: #0f76fc;
  padding: 0.5rem 1.5rem;
  font-size: 1em;
  cursor: pointer;
`;

const ProfileIcon = styled.i``;

const ToggleableMenu: React.FunctionComponent<PropTypes> = props => {
  const [expanded, setExpanded] = useState(false);
  const [showModal, setShowModal] = useState(false);
  const { loggedIn, iconName } = props;
  const getModalProps = () => {
    return {
      Embedded: Login,
      isOpen: showModal,
      closeModal: () => {
        setShowModal(!showModal);
      },
      parentProps: { ...props }
    };
  };
  useEffect(() => {
    const collapse = () => {
      setExpanded(false);
    };
    const main = document.getElementById("main");
    if (main) {
      main.addEventListener("click", collapse);
    }
    return () => {
      main ? main.removeEventListener("click", collapse) : null;
    };
  }, []);
  const handleToggleClick = (event: React.MouseEvent) => {
    setExpanded(!expanded);
  };
  if (loggedIn) {
    return (
      <DropDownWrapper>
        <ProfileIcon className={iconName} onClick={handleToggleClick} />
        <DropdownMenu expanded={expanded}>
          <LinkWrapper>
            <Link href="/user/profile">
              <StyledLink onClick={handleToggleClick}>Profile</StyledLink>
            </Link>
          </LinkWrapper>
          <LinkWrapper>
            <StyledLink href="/user/logout" onClick={handleToggleClick}>
              Logout
            </StyledLink>
          </LinkWrapper>
        </DropdownMenu>
      </DropDownWrapper>
    );
  } else {
    return (
      <DropDownWrapper>
        <ImportantLink
          onClick={() => {
            setShowModal(true);
          }}
        >
          Login
        </ImportantLink>
        {showModal && <Modal {...getModalProps()} />}
      </DropDownWrapper>
    );
  }
};

export default React.memo(ToggleableMenu);
