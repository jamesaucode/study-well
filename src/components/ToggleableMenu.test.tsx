/* eslint-disable */
import { mount } from "enzyme";
import ToggleableMenu from "./ToggleableMenu";

describe('Toggleable Menu', () => {
   it("Component renders", () => {
       const wrapper = mount(<ToggleableMenu />);
       expect(wrapper).toMatchSnapshot();
   }) 
})
