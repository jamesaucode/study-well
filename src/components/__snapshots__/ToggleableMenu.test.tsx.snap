// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggleable Menu Component renders 1`] = `
<ToggleableMenu>
  <ToggleableMenu__DropDownWrapper>
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "ToggleableMenu__DropDownWrapper-sc-1qfmg6u-0",
            "isStatic": true,
            "lastClassName": "OPYsp",
            "rules": Array [
              "position:relative;display:inline-block;align-items:center;padding:0 1rem;margin:0 auto;",
            ],
          },
          "displayName": "ToggleableMenu__DropDownWrapper",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "ToggleableMenu__DropDownWrapper-sc-1qfmg6u-0",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <div
        className="ToggleableMenu__DropDownWrapper-sc-1qfmg6u-0 OPYsp"
      >
        <ToggleableMenu__ImportantLink
          onClick={[Function]}
        >
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "ToggleableMenu__ImportantLink-sc-1qfmg6u-4",
                  "isStatic": true,
                  "lastClassName": "bVfOgW",
                  "rules": Array [
                    "font-size:1em;font-weight:600;padding:0 1rem;color:",
                    "#ffffff",
                    "cc;text-decoration:none;",
                    "border:1px solid #0f76fc;border-radius:3px;color:#0f76fc;padding:0.5rem 1.5rem;font-size:1em;cursor:pointer;",
                  ],
                },
                "displayName": "ToggleableMenu__ImportantLink",
                "foldedComponentIds": Array [
                  "ToggleableMenu__StyledLink-sc-1qfmg6u-3",
                ],
                "render": [Function],
                "styledComponentId": "ToggleableMenu__ImportantLink-sc-1qfmg6u-4",
                "target": "a",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            onClick={[Function]}
          >
            <a
              className="ToggleableMenu__StyledLink-sc-1qfmg6u-3 ToggleableMenu__ImportantLink-sc-1qfmg6u-4 bVfOgW"
              onClick={[Function]}
            >
              Login
            </a>
          </StyledComponent>
        </ToggleableMenu__ImportantLink>
      </div>
    </StyledComponent>
  </ToggleableMenu__DropDownWrapper>
</ToggleableMenu>
`;
